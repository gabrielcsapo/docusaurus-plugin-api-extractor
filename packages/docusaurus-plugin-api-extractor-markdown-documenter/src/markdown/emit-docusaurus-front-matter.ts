import visit from 'unist-util-visit';
import { Node, Literal, Parent } from 'unist';

interface HeadingNode extends Parent<Literal<string>> {
  depth: number;
}

/**
 * Emits the docusaurus frontmatter
 * @param id relative filepath
 * @returns
 */
export const emitDocusaurusFrontMatter = (id: string) => {
  return (tree: Node) => {
    let foundHeading = false;
    const parts = [`id: ${id}`, 'hide_title: true', 'custom_edit_url: null'];
    visit<HeadingNode>(tree, 'heading', (headingNode) => {
      if (
        !foundHeading &&
        headingNode.depth === 2 &&
        headingNode.children[0].type === 'text'
      ) {
        foundHeading = true;
        parts.push(`title: ${headingNode.children[0].value}`);
      }
    });

    if (id === 'index') {
      parts.push('slug: /');
    }

    (tree as Parent<Literal<string>>).children.unshift({
      type: 'yaml',
      value: parts.join('\n'),
    });
  };
};
