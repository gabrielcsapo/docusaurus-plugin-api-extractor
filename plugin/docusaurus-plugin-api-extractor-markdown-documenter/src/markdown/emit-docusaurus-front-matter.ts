import visit from 'unist-util-visit';
import { Node, Literal, Parent } from 'unist';

interface IHeadingNode extends Parent<Literal<string>> {
  depth: number;
}

/**
 * Emits the docusaurus frontmatter
 * @param id relative filepath
 * @returns
 */
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export const emitDocusaurusFrontMatter = (id: string) => {
  return (tree: Node) => {
    let foundHeading: boolean = false;
    const parts: string[] = [`id: ${id}`, 'hide_title: true', 'custom_edit_url: null'];
    visit<IHeadingNode>(tree, 'heading', (headingNode) => {
      if (!foundHeading && headingNode.depth === 2 && headingNode.children[0].type === 'text') {
        foundHeading = true;
        parts.push(`title: ${headingNode.children[0].value}`);
      }
    });

    if (id === 'index') {
      parts.push('slug: /');
    }

    (tree as Parent<Literal<string>>).children.unshift({
      type: 'yaml',
      value: parts.join('\n')
    });
  };
};
