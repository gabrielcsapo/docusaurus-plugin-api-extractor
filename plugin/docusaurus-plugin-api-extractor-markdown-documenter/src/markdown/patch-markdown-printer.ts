/**
 * mdast-util-to-markdown is what actually generates markdown in
 * remark. As part of this it will eagerly escape underscores when
 * they are part of "phrasing". This is problematic when a file path
 * for a link has underscores in it. When docusaurus validates links
 * it does a literal comparsion of the url and thus believes the
 * files don't exist.
 *
 * This is fixed in remark >= 14.0.0 however this requires that we use ESM.
 */

try {
  // eslint-disable-next-line @typescript-eslint/typedef
  const escapePatterns = require('mdast-util-to-markdown/lib/unsafe');

  // eslint-disable-next-line @typescript-eslint/typedef
  for (let i = escapePatterns.length - 1; i > 0; i--) {
    if (escapePatterns[i].character === '_') {
      escapePatterns.splice(i, 1);
    }
  }
} catch (e) {
  throw new Error(
    'Could not apply patch to markdown generate. Please open up an issue with the verion of this plugin'
  );
}
