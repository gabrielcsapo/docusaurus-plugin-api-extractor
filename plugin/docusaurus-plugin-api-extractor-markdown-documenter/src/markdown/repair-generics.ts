import visit from 'unist-util-visit';
import { Node, Data, Literal, Parent } from 'unist';

/**
 * This function is responsible for fixing how the hertiage types
 * (Implements, Extends, References) are emitted as markdown with
 * bold tags instead of the markdown equivalent followed by a
 * markdown link. This seems to be problematic for some
 * implementations of markdown and causes the link not to render
 * properly.
 * @returns {}
 */
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export const repairGenerics = () => {
  return (tree: Node<Data>) => {
    let inReturnSection: boolean = false;
    visit<Parent>(tree, 'paragraph', (paragraphNode) => {
      if (inReturnSection) {
        paragraphNode.children.forEach((child) => {
          if (child.type === 'text') {
            (child as Literal<string>).value = (child as Literal<string>).value
              .replace('<', '___REPAIR_LT___')
              .replace('>', '___REPAIR_GT___');
          }
        });
        inReturnSection = false;
      }

      if (paragraphNode.children.length === 3) {
        const [maybeBoldStart, maybeReturnsText, maybeBoldEnd] = paragraphNode.children as Literal<string>[];

        if (
          maybeBoldStart.value === '<b>' &&
          maybeReturnsText.value === 'Returns:' &&
          maybeBoldEnd.value === '</b>'
        ) {
          inReturnSection = true;
        }
      }
    });
  };
};
