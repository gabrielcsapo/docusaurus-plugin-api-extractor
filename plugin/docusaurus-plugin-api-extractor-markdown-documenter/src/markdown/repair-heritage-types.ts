import visit from 'unist-util-visit';
import { Node, Data, Literal, Parent } from 'unist';

interface IStrongNode extends Parent {
  type: 'strong';
}

/**
 * This function is responsible for fixing how the hertiage types
 * (Implements, Extends, References) are emitted as markdown with
 * bold tags instead of the markdown equivalent followed by a
 * markdown link. This seems to be problematic for some
 * implementations of markdown and causes the link not to render
 * properly.
 * @returns {}
 */
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export const repairHeritageTypes = () => {
  return (tree: Node<Data>) => {
    visit<Parent>(tree, 'paragraph', (paragraphNode) => {
      const strong: IStrongNode = {
        type: 'strong',
        children: []
      };

      visit<Literal<string>>(paragraphNode, 'html', (htmlNode) => {
        const index: number = paragraphNode.children.indexOf(htmlNode);
        const textNode: Literal<string> = paragraphNode.children[index + 1] as Literal<string>;

        if (
          htmlNode.value === '<b>' &&
          textNode.type === 'text' &&
          (textNode.value === 'Implements:' ||
            textNode.value === 'Extends:' ||
            textNode.value === 'References:')
        ) {
          const start: number = index + 1;

          let seenEndTag: boolean = false;
          paragraphNode.children.splice(index, 1, strong);

          for (let current: number = paragraphNode.children.length - 1; start <= current; current--) {
            const newChild: Literal<string> = paragraphNode.children[current] as Literal<string>;

            if (seenEndTag) {
              paragraphNode.children.splice(current, 1);
              strong.children.unshift(newChild);
            } else if (newChild.type === 'html' && newChild.value === '</b>') {
              seenEndTag = true;
              paragraphNode.children.splice(current, 1);
            }
          }
        }
      });
    });
  };
};
