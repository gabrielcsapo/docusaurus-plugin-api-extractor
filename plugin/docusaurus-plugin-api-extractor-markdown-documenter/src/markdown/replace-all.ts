// Polyfill for String.prototype.replaceAll
export const replaceAll = (str: string, find: string, replacement: string): string =>
  str.split(find).join(replacement);
