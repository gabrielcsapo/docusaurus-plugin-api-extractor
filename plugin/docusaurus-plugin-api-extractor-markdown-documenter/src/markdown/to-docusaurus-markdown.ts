import remark from 'remark';
import remarkFrontmatter from 'remark-frontmatter';
import { repairHeritageTypes } from './repair-heritage-types';
import { emitDocusaurusFrontMatter } from './emit-docusaurus-front-matter';
import './patch-markdown-printer';

/**
 * @param markdownString
 * @returns string
 */
export const toDocusaurusMarkDown = (markdownString: string, id: string): string => {
  return remark()
    .use(remarkFrontmatter, ['yaml'])
    .use(repairHeritageTypes)
    .use(emitDocusaurusFrontMatter, id)
    .processSync(markdownString).contents as string;
};
