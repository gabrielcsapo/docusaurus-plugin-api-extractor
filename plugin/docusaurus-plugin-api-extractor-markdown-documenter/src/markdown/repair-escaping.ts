import visit from 'unist-util-visit';
import { Node, Data, Literal, Parent } from 'unist';
import { replaceAll } from './replace-all';

/**
 * This function is responsible for fixing how the hertiage types
 * (Implements, Extends, References) are emitted as markdown with
 * bold tags instead of the markdown equivalent followed by a
 * markdown link. This seems to be problematic for some
 * implementations of markdown and causes the link not to render
 * properly.
 * @returns {}
 */
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export const repairEscaping = () => {
  return (tree: Node<Data>) => {
    visit<Parent>(tree, 'paragraph', (paragraphNode) => {
      paragraphNode.children.forEach((child) => {
        if (child.type === 'text') {
          let text: string = (child as Literal<string>).value;
          text = replaceAll(text, '<', '___REPAIR_LT___');
          text = replaceAll(text, '>', '___REPAIR_GT___');
          (child as Literal<string>).value = text;
        }
      });
    });
  };
};
