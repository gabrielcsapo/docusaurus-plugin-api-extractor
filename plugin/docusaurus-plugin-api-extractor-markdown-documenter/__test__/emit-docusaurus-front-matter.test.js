import remark from 'remark';
import { emitDocusaurusFrontMatter } from '../dist/markdown/emit-docusaurus-front-matter';
import remarkFrontmatter from 'remark-frontmatter';

it('Emits frontmatter with "A Class" title', () => {
  const out = remark()
    .use(remarkFrontmatter)
    .use(emitDocusaurusFrontMatter, 'a-class')
    .processSync('## A Class').contents;
  expect(out).toMatchSnapshot();
});

it('Emits frontmatter specific to index', () => {
  const out = remark()
    .use(remarkFrontmatter)
    .use(emitDocusaurusFrontMatter, 'index')
    .processSync('## Example').contents;
  expect(out).toMatchSnapshot();
});
