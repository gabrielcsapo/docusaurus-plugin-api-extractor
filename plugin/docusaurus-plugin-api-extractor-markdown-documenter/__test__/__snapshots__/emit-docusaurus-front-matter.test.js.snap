// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emits frontmatter specific to index 1`] = `
"---
id: index
hide_title: true
custom_edit_url: null
title: Example
slug: /
---

## Example
"
`;

exports[`Emits frontmatter with "A Class" title 1`] = `
"---
id: a-class
hide_title: true
custom_edit_url: null
title: A Class
---

## A Class
"
`;
